nvflare/edge/assessors/model_update.py

```diff
@@ -53,10 +56,154 @@ def __init__(
         self.model_manager = None
         self.device_manager = None
         self.max_model_version = max_model_version
+        self.device_wait_timeout = device_wait_timeout
         self.update_lock = threading.Lock()
-        self.start_time = None
+        self.device_wait_start_time = None
+        self.should_stop_job = False
+        self.timeout_future = None
+        self.thread_pool = ThreadPoolExecutor(max_workers=1, thread_name_prefix="DeviceTimeout")
+        self.timeout_wait_event = threading.Event()
+        self.timeout_thread_active = False
         self.register_event_handler(EventType.START_RUN, self._handle_start_run)
 
+    def _check_device_timeout(self, fl_ctx: FLContext) -> bool:
+        """Check if device wait timeout has been exceeded.
+
+        Args:
+            fl_ctx: FL context
+
+        Returns:
+            bool: True if timeout exceeded, False otherwise
+        """
+        if self.device_wait_start_time is None:
+            return False
+
+        if time.time() - self.device_wait_start_time > self.device_wait_timeout:
+            return True
+        return False
+
+    def _log_device_wait_status(self, fl_ctx: FLContext, message_prefix: str = ""):
+        """Log current device wait status with countdown information."""
+        if self.device_wait_start_time is not None:
+            current_time = time.time()
+            elapsed = current_time - self.device_wait_start_time
+
+            # Only log if we haven't logged recently (rate limiting)
+            if not hasattr(self, "_last_status_log_time"):
+                self._last_status_log_time = 0
+
+            # Adaptive logging frequency based on urgency
+            remaining_time = self.device_wait_timeout - elapsed
+            if remaining_time > 0:
+                # Determine logging interval based on remaining time
+                if remaining_time > 60:  # More than 1 minute: log every 30 seconds
+                    log_interval = 30.0
+                elif remaining_time > 30:  # 30 seconds to 1 minute: log every 15 seconds
+                    log_interval = 15.0
+                elif remaining_time > 10:  # 10 to 30 seconds: log every 5 seconds
+                    log_interval = 5.0
+                else:  # Final 10 seconds: log every 2 seconds
+                    log_interval = 2.0
```

Comment:
while adaptive might be good, should we instead let user decide the logging interval? I am neutral here, either way is OK